draft-gpew-priv-ppm.md

```diff
@@ -704,49 +716,39 @@ struct {
   TaskID task_id;
   opaque agg_param<0..2^16-1>;  // VDAF aggregation parameter
   opaque helper_state<0..2^16>; // helper's opaque state
-  AggregateSubReq seq<1..2^24-1>;
+  AggregateSub seq<1..2^24-1>;
 } AggregateReq;
 ~~~
 
+[TODO: Split into AggregateReq and AggregateInitReq so that only the latter
+carries the aggregation parameter.]
+
 The structure contains the PPM task, an opaque, VDAF-specific aggregation
 parameter, an opaque *helper state* string, and a sequence of *sub-requests*,
 each corresponding to a unique client report. Sub-requests are structured as
 follows:
 
 ~~~
 struct {
-  Nonce nonce;                     // Equal to Report.nonce.
-  Extension extensions<4..2^16-1>; // Equal to Report.extensions.
-  EncryptedInputShare helper_share;
+  Nonce nonce;               // The report nonce
   opaque message<0..2^16-1>; // VDAF message
-} AggregateSubReq;
+} AggregateSub;
 ~~~
 
-The `nonce` and `extensions` fields have the same value as those in the
-report uploaded by the client. Similarly, the `helper_share` field is the
-`EncryptedInputShare` from the `Report` whose index in
-`Report.encrypted_input_shares` is equal to the index of `[aggregator]` in the
-task's aggregator endpoints. [OPEN ISSUE: We usually only need to send this in
-the first aggregate request. Shall we exclude it in subsequent requests
-somehow?] The remainder of the structure is dedicated to VDAF-specific request
-parameters.
+The `nonce` field has the same value as those in the leader's report share
+uploaded by the client. The remainder of the structure is dedicated to
```

Comment:
Hmm, now that the nonce<>client uniqueness is no longer enforced by a single party, what happens if, say, two benign clients happen to generate colliding nonce values, and each aggregator picks a different one while filtering for anti-replay? We could maybe increase the width of the nonce random value to hedge against this collision, but it's not clear to me that's sufficient.

Reply:
Oh interesting! To be clear, before #169 we still had both aggregators enforce anti-replay. The difference is that we no longer of single party pick the set of reports that are processed. Nice catch.

This basically amounts to an attack on correctness. What will happen for prio3 or poplar1 is that the aggregators will derive different verification randomness, which will cause them to deem the report invalid and reject it (with high probability).

However, where this edge case *would* matter is if a VDAF were used that didn't have any verifiability. (We should probably not get in the habbit of calling such a scheme a VDAF: see https://github.com/cjpatton/vdaf/issues/20.) This was already requested in #45, so it's within the realm of possibility.

Note that a collision fairly likely, since the nonce only contains 64 random bits. (The remaining bits are a timestamp.) We may consider bumping this to 128 just to be on the safe side.

How about addressing this by leaving an OPEN ISSUE in the text? It would reference #45 and we would leave a note there to make sure we keep track of this.

Reply:
That's a fine resolution, yep. Let's also note that this type of collision will cause verification failures with overwhelming probability?

Reply:
Will do.